### PR TITLE
feat(firestore): include metadata property on snapshot (#567)

### PR DESCRIPTION
**Pull request by phjardas**
I've created the Android implementation based on your proposed interface. I've removed the `source` option because it is only supported on the web platform and does not exist in the Android SDK.

### CODE REVIEW

File: packages/firestore/src/definitions.ts [hunk 1]
```diff
@@ -377,14 +377,6 @@ export interface SnapshotListenerOptions {
...
    */
   readonly includeMetadataChanges?: boolean;
-  /**
-   * Set the source the query listens to.
-   * The source `default` listens to both cache and server.
-   *
-   * @since 6.2.0
-   * @default "default"
-   */
-  readonly source?: 'default' | 'cache';
```
Comment (robingenz) [inline packages/firestore/src/definitions.ts:387]: Please revert this change. It's totally fine if this property is only supported on web. Just add a comment `Only available for Web.`.

File: packages/firestore/src/web.ts [hunk 2]
```diff
@@ -230,10 +230,7 @@ export class FirebaseFirestoreWeb
     const firestore = getFirestore();
     const unsubscribe = onSnapshot(
       doc(firestore, options.reference),
-      {
-        includeMetadataChanges: options.includeMetadataChanges,
-        source: options.source,
-      },
```
Comment (robingenz) [inline packages/firestore/src/web.ts:236]: Please revert this.

File: packages/firestore/src/web.ts [hunk 3]
```diff
@@ -268,10 +265,7 @@ export class FirebaseFirestoreWeb
     );
     const unsubscribe = onSnapshot(
       collectionQuery,
-      {
-        includeMetadataChanges: options.includeMetadataChanges,
-        source: options.source,
-      },
```
Comment (robingenz) [inline packages/firestore/src/web.ts:274]: Please revert this.

File: packages/firestore/src/web.ts [hunk 4]
```diff
@@ -305,10 +299,7 @@ export class FirebaseFirestoreWeb
...
```
Comment (robingenz) [inline packages/firestore/src/web.ts:311]: Please revert this.

File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/FirebaseFirestoreHelper.java [hunk 5]
```diff
@@ -126,4 +128,11 @@ private static JSArray createJSArrayFromArrayList(ArrayList arrayList) {
         }
         return array;
     }
+
+    public static JSObject createSnapshotMetadata(DocumentSnapshot snapshot) {
```
Comment (robingenz) [inline packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/FirebaseFirestoreHelper.java:132]:
```suggestion
    public static JSObject createSnapshotMetadataResult(DocumentSnapshot snapshot) {
```

**Review by robingenz — CHANGES_REQUESTED** [review]
You also need to make changes to the `FirebaseFirestorePlugin.java` file.

**Comment by phjardas** [comment]
Hope everything is right now. :-)

**Comment by phjardas** [comment]
I'll have a go at the iOS implementation this afternoon.

File: packages/firestore/src/definitions.ts [hunk 6]
```diff
@@ -377,10 +377,13 @@ export interface SnapshotListenerOptions {
    * @default false
    */
   readonly includeMetadataChanges?: boolean;
+
```
Comment (robingenz) [inline packages/firestore/src/definitions.ts:380]: Please remove this empty line.

File: packages/firestore/src/definitions.ts [hunk 7]
```diff
@@ -377,10 +377,13 @@ export interface SnapshotListenerOptions {
...
   /**
    * Set the source the query listens to.
    * The source `default` listens to both cache and server.
    *
+   * Note that this option is only available on the web. It is ignored on Android or iOS.
```
Comment (robingenz) [inline packages/firestore/src/definitions.ts:385]:
```suggestion
   * Only available for Web.
```

File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddDocumentSnapshotListenerOptions.java [hunk 8]
```diff
@@ -1,19 +1,31 @@
...
     private String callbackId;
 
-    public AddDocumentSnapshotListenerOptions(String reference, String callbackId) {
+    public AddDocumentSnapshotListenerOptions(
+            String reference,
+            @Nullable Boolean includeMetadataChanges,
+            String callbackId
+    ) {
         this.reference = reference;
+        this.includeMetadataChanges = includeMetadataChanges != null ? includeMetadataChanges : false;
```
Comment (robingenz) [inline packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddDocumentSnapshotListenerOptions.java:17]:
```suggestion
        this.includeMetadataChanges = includeMetadataChanges == null ? false : includeMetadataChanges;
```

Keep it simple.

File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddCollectionSnapshotListenerOptions.java [hunk 9]
```diff
@@ -23,15 +23,19 @@ public class AddCollectionSnapshotListenerOptions {
...
         String reference,
         @Nullable JSObject compositeFilter,
         @Nullable JSArray queryConstraints,
+        @Nullable Boolean includeMetadataChanges,
         String callbackId
     ) throws JSONException {
         this.reference = reference;
         this.compositeFilter = FirebaseFirestoreHelper.createQueryCompositeFilterConstraintFromJSObject(compositeFilter);
         this.queryConstraints = FirebaseFirestoreHelper.createQueryNonFilterConstraintArrayFromJSArray(queryConstraints);
+        this.includeMetadataChanges = includeMetadataChanges != null ? includeMetadataChanges : false;
```
Comment (robingenz) [inline packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddCollectionSnapshotListenerOptions.java:38]:
```suggestion
        this.includeMetadataChanges = includeMetadataChanges == null ? false : includeMetadataChanges;
```

Keep it simple.

File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddCollectionGroupSnapshotListenerOptions.java [hunk 10]
```diff
@@ -22,15 +25,19 @@ public class AddCollectionGroupSnapshotListenerOptions {
...
+            String reference,
+            @Nullable JSObject compositeFilter,
+            @Nullable JSArray queryConstraints,
+            @Nullable Boolean includeMetadataChanges,
+            String callbackId
     ) throws JSONException {
         this.reference = reference;
         this.compositeFilter = FirebaseFirestoreHelper.createQueryCompositeFilterConstraintFromJSObject(compositeFilter);
         this.queryConstraints = FirebaseFirestoreHelper.createQueryNonFilterConstraintArrayFromJSArray(queryConstraints);
+        this.includeMetadataChanges = includeMetadataChanges != null ? includeMetadataChanges : false;
```
Comment (robingenz) [inline packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddCollectionGroupSnapshotListenerOptions.java:40]:
```suggestion
        this.includeMetadataChanges = includeMetadataChanges == null ? false : includeMetadataChanges;
```

Keep it simple.

**Review by robingenz — CHANGES_REQUESTED** [review]

File: packages/firestore/src/definitions.ts (hunk 7 above)
Comment (phjardas) [reply packages/firestore/src/definitions.ts]: I've kept the "ignored on Android and iOS" part.

File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddDocumentSnapshotListenerOptions.java (hunk 8 above)
Comment (phjardas) [reply packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddDocumentSnapshotListenerOptions.java]: I guess it's a matter of taste. I prefer to keep the default at the end, analogous to the JS syntax `includeMetadataChanges ?? false`. However, it's your project, so I'll stick to your coding style. 😊

File: packages/firestore/src/definitions.ts (hunk 7 above)
Comment (robingenz) [reply packages/firestore/src/definitions.ts]: Please adopt my suggestion exactly. This text is used uniformly across all plugins.

File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddDocumentSnapshotListenerOptions.java (hunk 8 above)
Comment (robingenz) [reply packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/AddDocumentSnapshotListenerOptions.java]: You're right. It really is a matter of taste. Thank you for changing it anyway.

**Comment by robingenz** [comment]
Please run `npm run fmt` to format your code.

**Comment by robingenz** [comment]
@phjardas You want to push the ios implementation on this branch too?

**Comment by phjardas** [comment]
> @phjardas LGTM! You want to push the ios implementation on this branch too?

Yeah, let's hope I find time tomorrow. I've never worked with Swift or iOS before. 😬

**Comment by robingenz** [comment]
Let me know if you have any questions. ;)

**Comment by phjardas** [comment]
I just pushed the iOS implementation. It does build and verify, but I don't know how to actually test it.

**Review by robingenz — CHANGES_REQUESTED** [review]
It seems like the actual iOS implementation in `FirebaseFirestore.swift` is missing.

**Comment by phjardas** [comment]
Oops. 🙈 I've added the missing code. I still don't understand how I could actually test this plugin with my real capacitor app to verify everything's working now. Can you give me a pointer?

**Comment by robingenz** [comment]
There are several ways to do this, the easiest is probably to simply use the pre-release, see https://github.com/capawesome-team/capacitor-firebase/runs/29462389318

```
npm i https://pkg.pr.new/capawesome-team/capacitor-firebase/@capacitor-firebase/firestore@84be8d3
```

**Comment by robingenz** [comment]
Let me know as soon as you have tested it.

**Review by robingenz — APPROVED** [review]

**Comment by phjardas** [comment]
✅ Tested successfully in a live app on Web, Android and iOS.

**Comment by phjardas** [comment]
BTW: this plugin and feature are used in rescuetablet.com ;-)

**Comment by robingenz** [comment]
That's awesome! 😍 Thank you for sharing! If you ever plan to deliver [Live Updates](https://capawesome.io/blog/how-live-updates-for-capacitor-work/), I would love to welcome you to [Capawesome Cloud](https://capawesome.io/cloud/live-updates/) (the only live update solution that is **EU based & hosted**). 😆

**Comment by phjardas** [comment]
Could you be so kind to soon make a patch release with this feature? 🙏🏻

We're currently using CapGo, and I'd rather not change the provider at this time. Unless you have very compelling arguments. 😜

**Comment by robingenz** [comment]
> Could you be so kind to soon make a patch release with this feature? 🙏🏻

Since all plugins are always updated at the same time, we wait until there are enough changes. Otherwise we would publish updates for 11 plugins for each change. ^^ I recommend that you use the pre-release until version 6.2.0 is published.

> We're currently using CapGo, and I'd rather not change the provider at this time. Unless you have very compelling arguments. 😜

Feel free to reach out to me on LinkedIn. Maybe I can convince you. 😉
